src/vm/amd64/cgencpu.h

```diff
@@ -110,7 +110,7 @@ EXTERN_C void FastCallFinalizeWorker(Object *obj, PCODE funcPtr);
 // as large as the largest FieldMarshaler subclass. This requirement
 // is guarded by an assert.
 //=======================================================================
-#define MAXFIELDMARSHALERSIZE                   40
+#define MAXFIELDMARSHALERSIZE                   48
```

Comment:
These structures are allocated eagerly today by the type loader for nearly every struct and some classes too, even if they have nothing to do with interop. Is it worth it to allocate 8 bytes more per field just for this cleanup? It does not sound very pay-for-play.

I know that none of this particularly efficient today, but it would be useful to not make it even less efficient.

Reply:
That is a good point - I didn't realize they were allocated so early. It is also interesting that the size increases by 8 bytes with just an enum increase - considering the enum defaults to 32-bits. I am going to assume this is because of some padding requirement.

@jkoritzinsky This change might get a bit more complicated, but I think it might be useful to do some investigation on the layout of the `FieldMarshaler` classes and try to determine if we can use bits that are currently unused in the current layout.

Reply:
> the size increases by 8 bytes with just an additional enum

That's because of alignment.

Refactoring to make FieldMarshalers allocated on demand would help address my concern too...

Reply:
>That's because of alignment.

Yeah, there might be other padding we can use in this struct. Alternatively, we could use templates for this and sacrifice some code section of the binary in lue of making local members? This could be done for both `NStructFieldType` and `NativeFieldCategory`. Unsure how much code churn that would be.

Yeah I am not sure how easy that would be to change this to allocate on demand. There are a lot of levels here.

@jkoritzinsky The allocation in question is here: https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L12148-L12149

Reply:
It would be a lot more work to make it load on demand.

One thing we could do is change the `NStructFieldType`, `NativeFieldSubcategory`, and `NativeFieldCategory` enumerations only 2-bytes wide each (which would bring the size back down to what it was before. There are a relatively small number of NFT_* types, so limiting that enum value to 16-bits wouldn't be a problem at all. And if we don't want to make the enums 2-bytes wide each, we could just use the higher bits of the `m_nft` field for the categories instead of adding a new field.

Reply:
I've made the size of each of the enums only 2-bytes wide. So, the size of the largest `FieldMarshaler` has dropped back down to what it was before.